### PR TITLE
Implement better configure-dirty checks.

### LINKED ISSUE #385: "No build system was generated yet" errors from CMake Server
It looks like we try to get the codemodel from CMake Server before the generation is finished. This is the most common rollbar issue ATM, even though it doesn't seem to cause any real trouble?

### PR DESCRIPTION
**Pull request by vector-of-bool**
CMake's dirty signal is a bit unreliable, but it does let us query all files
that are inputs to the build. We can use this information to implement
our own dirty checks.

This commit also tweaks some auto-saving so that we can be sure that
configure output is written to the terminal at the proper time. It also
moves auto-saving out of the driver logic into the business logic code.

[Fix #385], possibly?

### CODE REVIEW

File: src/cms-driver.ts [hunk 1]
```diff
@@ -18,6 +18,20 @@ import * as util from './util';
 
 const log = createLogger('cms-driver');
 
+class CMakeInputFile {
```
Comment (KoeMai) [inline src/cms-driver.ts]: Please move the cmake dirty check into a separate file, it is an encapsulated unit.

File: src/cmake-tools.ts [hunk 2]
```diff
@@ -484,8 +514,13 @@ export class CMakeTools implements vscode.Disposable, api.CMakeToolsAPI {
    * Implementation of `cmake.build`
    */
   async build(target_?: string): Promise<number> {
-    // First, reconfigure if necessary
-    if (await this._needsReconfigure()) {
+    // First, save open files
+    if (!await this.maybeAutoSaveAll()) {
+      return -1;
```
Comment (KoeMai) [inline src/cmake-tools.ts]: Is this error code unique and is it needed?

Comment (vector-of-bool) [reply src/cmake-tools.ts]: Not unique, but its purpose is to signal to the caller that the build "failed" (because saving files failed).

**Comment by KoeMai** [comment]
Please add an unit test for file change detection, dirty generation and cmake server cmakeInputs.

**Comment by vector-of-bool** [comment]
Want to give it another look? I'm almost ready to merge, and it will seal up the things I want for 0.11.1

**Pull request merged**